### PR TITLE
Improve phase assembly errors

### PR DESCRIPTION
**Pull request by som-snytt**
This was residual to the minimal PR, so it can be skipped unless there is a complaint to follow up on.

`-Werror` on bad phase config. Include core phase names when checking for typos in constraints.

Previous description that pertains to commits merged previously:

Don't issue "phase" warnings under Scaladoc, to minimize build noise. There is little peril in not warning, since errors in plugin configuration are likely to be seen already when compiling.

The motivating use case is that a typo in a "runsBefore pickler" constraint for Scala-js was corrected but warned anyway in Scaladoc, which has no pickler phase.

Previously to this follow-up PR, "better monadic for" would warn for "runsBefore patmat". This PR reverts the fix of adding mock components to Scaladoc just for these constraints. Instead, Scaladoc does not warn by default if a constraint does not name a component. (It will warn under `-Vdebug`.)

This is deemed useful and benign for the use cases described: a plugin "runsBefore" a compiler phase that does not exist in Scaladoc. The compiler warns and offers spelling advice.

Both tools warn if a plugin component is not installed because it is not "reachable" in the graph of phase constraints.

Note that is feasible for a component to have no "runsAfter" constraint, if another component "runsBefore" it.

The net difference with 2.13.14 is that the compiler will never silently omit a plugin phase. This PR follows 2.13.14 in being more circumspect about bad "runsBefore" constraints, but also warns about typos. 

Follows up commits merged for https://github.com/scala/bug/issues/13028

### CODE REVIEW

File: test/scaladoc/run/t8755c/ploogin_1.scala [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  import global._
+
+  val name = "ploogin"
+  val description = "A sample plugin for testing."
+  val components = List[PluginComponent](TestComponent)
+
+  private object TestComponent extends PluginComponent {
+    val global: Ploogin.this.global.type = Ploogin.this.global
+    override val runsBefore = Nil
+    val runsAfter = List("refchecks")
```
Comment (som-snytt) [inline test/scaladoc/run/t8755c/ploogin_1.scala]: other way around. So many combinations. Where is Scalacheck to fuzz your test source when you need it?

**Comment by som-snytt** [comment]
I forgot to say last-only.

File: src/partest/scala/tools/partest/DirectTest.scala [hunk 2]
```diff
@@ -53,12 +53,15 @@ abstract class DirectTest {
...
-  def newSettings(args: List[String]) = (new Settings).tap { s =>
-    val allArgs = args ++ tokenize(debugSettings)
+  def newSettings(args: List[String]): Settings = newSettings().tap { s =>
+    val allArgs = debugSettings.pipe(db => if (db.isEmpty) args else args ++ tokenize(db))
     log(s"newSettings: allArgs = $allArgs")
     val (success, residual) = s.processArguments(allArgs, processAll = false)
     assert(success && residual.isEmpty, s"Bad settings [${args.mkString(",")}], residual [${residual.mkString(",")}]")
   }
+  // scaladoc has custom settings
+  def newSettings(): Settings = new Settings
```
Comment (sjrd) [inline src/partest/scala/tools/partest/DirectTest.scala]: Call this `newSettingsBase()` or something like that? With the same name as the overload I would expect this method to be an alias for `newSettings(args)` with some default `args`; not a base method that `newSettings(args)` *must* call to create its instance of `Settings`.

File: test/scaladoc/run/t8755/ploogin_1.scala [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+class Ploogin(val global: Global) extends Plugin {
+  import global._
+
+  val name = "ploogin"
+  val description = "A sample plugin for testing."
+  val components = List[PluginComponent](TestComponent)
+
+  private object TestComponent extends PluginComponent {
+    val global: Ploogin.this.global.type = Ploogin.this.global
+    //override val runsBefore = List("jvm")
```
Comment (sjrd) [inline test/scaladoc/run/t8755/ploogin_1.scala]: Is the commented out form intentional? Or did you want to remove the line entirely?

**Review by sjrd — APPROVED** [review]

File: test/scaladoc/run/t8755/ploogin_1.scala (hunk 3 above)
Comment (som-snytt) [reply test/scaladoc/run/t8755/ploogin_1.scala]: The three tests are combinations of runsBefore and runsAfter. Which is the override? I need a model to copy. In fact, I commented that I didn't even get that right on the third test and need a test generator.

The other question is: why is runsBefore deemed a second class constraint that has an empty default?

File: src/compiler/scala/tools/nsc/PhaseAssembly.scala [hunk 4]
```diff
@@ -249,11 +249,12 @@ object DependencyGraph {
           graph.addEdge(after, p.phaseName, Follows)
         for (before <- p.runsBefore if before.nonEmpty && checkConstraint(before, "runsBefore"))
           graph.addEdge(p.phaseName, before, Follows)
+        def noPhase(s: String): Boolean = s.isEmpty || !graph.components.contains(s)
         if (p != start && p != end)
-          if (p.runsRightAfter.forall(_.isEmpty) && p.runsAfter.forall(_.isEmpty))
+          if (p.runsRightAfter.forall(noPhase) && p.runsAfter.forall(noPhase))
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/PhaseAssembly.scala]: I'm unsure this is the right solution. Such a lax contraint allows the phase to end up anywhere. The algorithm is deterministic, but the result depends on the phase name: `a-plugin` gives `parser - a-plugin - namer ...`, but `plugin` gives `parser - namer - packageobjects - typer - plugin ....`.

In the case of Scaladoc, this means that a plugin with `runsAfter pickler`, for example, could end up before typer and would likely crash.

2.13.14 drops phases where either the `runsAfter` or the `runsRightAfter` constraint is invalid (with a warning). The phase is also dropped when both `runsAfter` and `runsRightAfter` are empty (silently).

2.13.14 silently ignores invalid the `runsBefore` constraints. That's why the `"pickle"` typo in Scala.js didn't matter (https://github.com/scala-js/scala-js/commit/aa84f641c653fbbfc1b14378329cdf89cf1c5564).

So overall I suggest
  - drop phases with invalid `runsAfter`/`runsRightAfter` constraints, warn (also in scaladoc: if someone writes `runsAfter pickler` they should get notified that the phase is dropped in scaladoc. since we have that warning in 2.13.14 already, it's not a breaking change. maybe there should be a way for a phase to declare whether it's intended to also run in scaladoc, but that can be a follow-up.)
  - ignore invalid `runsBefore` constraints, warn unless in scaladoc. that is already done in this PR.

wdyt?

Comment (sjrd) [reply src/compiler/scala/tools/nsc/PhaseAssembly.scala]: > * drop phases with invalid `runsAfter`/`runsRightAfter` constraints, warn (also in scaladoc: if someone writes `runsAfter pickler` they should get notified that the phase is dropped in scaladoc. since we have that warning in 2.13.14 already, it's not a breaking change. maybe there should be a way for a phase to declare whether it's intended to also run in scaladoc, but that can be a follow-up.)

Won't that warn for any phase that declares otherwise valid constraints on later phases? Like
```scala
    override val runsAfter = List("specialize")
    override val runsBefore = List("explicitouter")
```
We have several phases like that in Scala.js, and I'm sure there are plenty of those in other plugins. Those should be dropped without warning. They are in 2.13.14.

Or actually give the plugins a way to declare that they should run in Scaladoc, and then only warn for those if they get dropped. Otherwise there will be lots of false positives.

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/PhaseAssembly.scala]: OK I've deleted the long reply I wrote while my oatmeal burned on the stove.

I'll do lrytz's suggestion with special dispensation for scalac phase names: IIUC, warn for unknown constraints, otherwise silently ignore phase that runsAfter a phase not in the run. (I'll check if a warning here is noisy.)

I'll check scala-js under 2.13.14 and head; I checked using Seth's instructions yesterday but did not check `-Vphases` or any output.

The principle is not to be "more helpful" but not to change anyone's perceived DX.

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/PhaseAssembly.scala]: Just to add, I liked the simplicity of this edit.

Comment (lrytz) [reply src/compiler/scala/tools/nsc/PhaseAssembly.scala]: > We have several phases like that in Scala.js, and I'm sure there are plenty of those in other plugins. Those should be dropped without warning. They are in 2.13.14.

OK - in my testing, 2.13.14 scaladoc warns on invalid `runsAfter` constraints:

```
➜ sandbox java -cp $(cs fetch -p org.scala-lang:scala-compiler:2.13.14):. scala.tools.nsc.ScalaDoc -usejavacp -Xplugin:plugin.jar T.scala -Vdebug
warning: dropping dependency on node with no phase object: specialize
[running phase parser on T.scala]
[running phase namer on T.scala]
[running phase packageobjects on T.scala]
[running phase typer on T.scala]
```

But you're probably right, the safest way forward is to never warn in Scaladoc.

Comment (lrytz) [reply src/compiler/scala/tools/nsc/PhaseAssembly.scala]: > lrytz's suggestion with special dispensation for scalac phase names

not sure which suggestion you mean; I like the current PR

Comment (lrytz) [reply src/compiler/scala/tools/nsc/PhaseAssembly.scala]: so we'd just have to drop phases with invalid `runsAfter` constraints, instead of putting `runsAfter = parser`

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/PhaseAssembly.scala:254]: https://fosstodon.org/@som_snytt/113080518511283717

I intended to confirm that running after a phantom phase is just dropped, but that is my current IIUC. My old comment would have been, No, constraints must be verified. If you intend a phase not to run, then don't configure it; or, the component can report that it won't be running because there is no phase X. But practically, it's OK to say "runs after" means "runs after X if X runs at all".

The lrytz suggestion I meant was to warn always ("there is no pickler in scaladoc") vs just for typos ("picklers").

**Comment by som-snytt** [comment]
I'll take another look when my eyes are fresh, but some notes:

- I didn't fail outright on bad or missing "after", because "before erasure" seems fine in general (but then it should just say "after parser" explicitly)
- I was wondering why adding "after parser" to everything does not work: the distances work out the same, but the code for sorting at the same "level" checks for "incoming edge from previous level" (instead of "which phases at this level are runsRightAfter"), the anomaly was plugin between packageobjects and typer
- as lrytz noted, the existing "just make it follow parser" resulted in sorting by name of the "anchors" (`namer` is anchor of `namer -> packageobjects -> typer`)
- validate now checks reachable, which is not guaranteed by "after" given cycles; for edge cases "before parser" or cycles, this warns but should fail
- I tired of uncommenting the local debugging code
- the unit test uses a large phase set, so linear search of names is slow

File: src/compiler/scala/tools/nsc/PhaseAssembly.scala [hunk 5]
```diff
@@ -224,38 +234,92 @@ object DependencyGraph {
...
+            compilerPhases.filter(util.EditDistance.levenshtein(name, _) < 3) match {
+              case Nil => ""
+              case close => s" - did you mean ${util.StringUtil.oxford(close, "or")}?"
+            }
+          warnings.addOne(s"No phase `$name` for ${p.phaseName}.$constraint$help")
+          badConstraint = true
+        }
       }
+      // add edges for each constraint
+      for (after <- p.runsRightAfter if after.nonEmpty && checkConstraint(after, "runsRightAfter", warn=true)) {
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/PhaseAssembly.scala]: I found one more corner case... 2.13.14 initially adds the `platformPhases` when running Scaladoc, then they are silently dropped when the algorithm runs. But that affects the behavior for a plugin phase with `runsAfter flatten`: it's dropped without warning.

There's also Sebastien's report that 2.13.14 Scaladoc drops Scala.js phases without warning (https://github.com/scala/scala/pull/10856#discussion_r1744224563).

I assume the warning in 2.13.14 for dropping a phase is buggy / not always issued when it "should" have been.

So should we do `warn = warnAll` here?

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/PhaseAssembly.scala]: I remember that the weird construction of phases was why I wanted to "nail it down".

For the warning, I think we can design the behavior. My basis for not warning is that people have arbitrary plugins enabled for scalac in their build, and scaladoc (as a doc task) just inherits that cruft.

This makes me think that scaladoc should just be a plugin.

I haven't done another pass on the PR yet, but oh I see you're saying don't warn here under scaladoc. I bet you're right.

**Comment by som-snytt** [comment]
appreciating the blue notation for updating a check file.

image

**Comment by som-snytt** [comment]
```
         typer   4  the meat and potatoes: type the trees
            C1   0  C1 tests phase assembly
            C2   0  C2 tests phase assembly
superaccessors   7  add super accessors in traits and nested classes
```
but `-Vprint:5` works

**Comment by som-snytt** [comment]
crashes 2.13.14 creating the dot file
```
    override val runsRightAfter = None
    override val runsAfter = List("typer")
    override val runsBefore = List("typer")
```
Realized I don't need N plugins but one plugin with N components to test the permutations.

In Knuth, "boolean basics" says here's a table of operations, there are 16 of them. I was like what? so in addition to not assuming I know anything, I assume I know less than I thought I did.

Edit: well, this PR also throws for reporting.
```
error: scala.reflect.internal.FatalError: Phases form a cycle: typer -> superaccessors -> extmethods -> pickler -> refchecks -> patmat -> C2 -> typer
```
for
```
    override val runsRightAfter = Option("patmat")
    override val runsAfter = List("typer")
    override val runsBefore = List("typer")
```
where I think we don't want to ignore a before constraint that correctly names a phase but the constraint can't be satisfied. (The original ticket 8755 mentions this behavior of runsRightAfter.)

Naturally it also detects the other case
```
error: scala.reflect.internal.FatalError: Phases form a cycle: typer -> C4 -> typer
```

I think I already knew 2.13.14 scalac fails on these obvious constraints.
```
    override val runsRightAfter = None
    override val runsAfter = List("typer")
    override val runsBefore = List("erasure")

error: scala.reflect.internal.FatalError: Phase erasure can't follow explicitouter, created phase-order.dot
```
So we don't want to remain bug-compatible, but merely convenience-compatible for Scaladoc.

Worth adding that on linked ticket 7905 ten years ago, I'd already noticed that generating the phase graph file throws.

The OG also complains that before constraint is not sufficient. This PR accepts C7 with no after constraint (only before terminal) because it is reachable through C8. I'm about to change that and preserve that one bit of buggy behavior, because Lukas said to, and because it is easy to describe: a runs-after constraint is required. TODO that's incorrect, 2.13.14 has the same behavior; scaladoc says `warning: dropping dependency on node with no phase object: patmat` but does not drop the phase (which runs). Maybe I misunderstood Lukas's comment, as I haven't tried it out yet.
```
         typer   4  the meat and potatoes: type the trees
            C1   0  C1 tests phase assembly
            C8   0  C8 is before C7 which specifies no after
superaccessors   7  add super accessors in traits and nested classes
            C7   0  C7 tests phase assembly if only a before constraint
    extmethods   9  add extension methods for inline classes
```

**Comment by SethTisue** [comment]
I feel bad even suggesting this, but perhaps it would be better to revert the original PR entirely, in order not to be under time pressure, then try to land a revised version in 2.13.16.

**Comment by som-snytt** [comment]
@SethTisue I see your point, but it's just shaking out a few edge cases. That's not going to change by waiting.

I'll add the comprehensive test for review of *all possible behaviors*.

However, it does belie the current practice that "it doesn't matter if we merge PRs because nothing will get tested until the week before the release." "Cramming" is what makes Scala a so-called "academic language", I just realized.

**Comment by som-snytt** [comment]
I am the last to understand that scala-js exposes different components for scaladoc:
```
    if (global.isInstanceOf[doc.ScaladocGlobal]) {
      List[NscPluginComponent](PrepInteropComponent)
    } else {
      List[NscPluginComponent](PreTyperComponentComponent, PrepInteropComponent,
          ExplicitInnerJSComponent, ExplicitLocalJSComponent, GenCodeComponent)
    }
```
so there is no mysterious convenience in trimming them. I haven't refreshed my memory about backend phases yet.

**Comment by SethTisue** [comment]
It's a good counterargument to observe that the only additional plugins, not in the community build, that are likely to get tested with this, and only at release time, are Scala.js, Scala Native, and my own Fortify plugin.

**Comment by som-snytt** [comment]
I'm just going to finish Jane the Virgin before I push another commit, but I realize Scala 2 is basically EOL so you have your priorities, it's really no skin off my nose. Edit: that was followed by a Charmed marathon.

**Comment by sjrd** [comment]
> I am the last to understand that scala-js exposes different components for scaladoc:

Huh. I had completely forgotten about that. Sorry for leading the instigation in the wrong direction. 🙁

**Comment by som-snytt** [comment]
Celebrating 11 yrs since scala-js plugin.

File: src/compiler/scala/tools/nsc/PhaseAssembly.scala [hunk 6]
```diff
@@ -224,38 +234,92 @@ object DependencyGraph {
...
+      }
+      for (after <- p.runsAfter if after.nonEmpty && !p.runsRightAfter.contains(after) && checkConstraint(after, "runsAfter", warn=warnAll)) {
+        addConstraint(after, p.phaseName, Follows)
+        constrained = true
+      }
+      for (before <- p.runsBefore if before.nonEmpty && checkConstraint(before, "runsBefore", warn=warnAll)) {
+        addConstraint(p.phaseName, before, Follows)
+        constrained = true
+      }
+      // if it doesn't follow a phase in the run, replace this component with a no-op that follows start
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/PhaseAssembly.scala]: If there are two phases where `p2 runsAfter p1` and `p1 runsAfter typo`, then only `p1` is replaced by a noop.

Here's another approach, I'll clean it up and adjust / add tests: https://github.com/scala/scala/compare/2.13.x...lrytz:scala:pr10856?expand=1

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/PhaseAssembly.scala]: I changed the noop to a list of names to filter out.

Comment (lrytz) [reply src/compiler/scala/tools/nsc/PhaseAssembly.scala]: Is that a change that you didn't push yet?

**Comment by lrytz** [comment]
`-Werror` doesn't work because the warnings are issued early, then `Run.compileUnits` resets the reporter (and its warning count)

**Comment by som-snytt** [comment]
Threw in a `-Werror` check, similar to the check at `compileSources`. "Something went wrong so summarize now."

I did not yet add the test for "two components are dropped".

I intended to comment in the minor refactor in PhaseAssembly.traverse that filtering for `distance >= 0` eliminates the unreachable (dropped) component(s). 

I could polish more, as there are a few spots where I can't see my face.

File: test/scaladoc/run/t8755b/ploogin_1.scala [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  import global._
+
+  val name = "ploogin"
+  val description = "A sample plugin for testing."
+  val components = List[PluginComponent](TestComponent)
+
+  private object TestComponent extends PluginComponent {
+    val global: Ploogin.this.global.type = Ploogin.this.global
+    override val runsBefore = List("jvm")
+    val runsAfter = List("typer", "refchecks")
```
Comment (lrytz) [inline test/scaladoc/run/t8755b/ploogin_1.scala]: I see, https://github.com/scala/scala/pull/10856 drops the phase because one of the `runsAfter` constraints is invalid. I think either way is fine here.

2.13.14 scaladoc drops the "refchecks" constraint and runs the phase after "typer".

**Comment by lrytz** [comment]
@som-snytt in the interest of moving 2.13.15 ahead I'd like to focus on what we need for that, so I submitted a minimal diff (https://github.com/scala/scala/pull/10858). Can we go ahead with this for now?

The remaining changes from my PR to this one: https://github.com/lrytz/scala/compare/pr10856light...lrytz:scala:pr10856?expand=1

What we can address after 2.13.15
  - detect duplicate phase names. weirdly, `SubComponent` overrides `equals` to compare the phase name, so phases with the same name don't end up in `Global.phasesSet`
  - `-Werror` for phase assembly warnings

Can you summarize what else is in the remaining diff? Performance improvements, debug logging, code cleanups.

File: test/scaladoc/run/t8755b/ploogin_1.scala (hunk 7 above)
Comment (som-snytt) [reply test/scaladoc/run/t8755b/ploogin_1.scala]: Yes I should have documented it: if runsAfter means "depends on", then all those phases are required, and not "at least one".

**Comment by som-snytt** [comment]
On duplicate names, I mention somewhere that Scala 3 uses a fixed "spine" of phases which supports name duplication. One could explore how options should work that rely on names. Scala 3 options don't use phase id.

Also phase replacement was potentially a feature.

File: src/compiler/scala/tools/nsc/PhaseAssembly.scala [hunk 8]
```diff
@@ -226,50 +232,92 @@ object DependencyGraph {
...
+    var dropped = Set.empty[String]
+    val constraints = ListBuffer.empty[(String, String, Weight)]
+    def addConstraint(from: String, to: String, weight: Weight): Unit = constraints.addOne((from, to, weight))
+    def checkConstraint(p: SubComponent)(name: String, constraint: String, warn: Boolean): Boolean = isPhaseName(name) || {
+      if (warn) {
+        val knownNames = phaseNames ++ compilerPhases
+        val help = if (knownNames.contains(name)) "" else
+          knownNames.filter(util.EditDistance.levenshtein(name, _) < 3).toList match {
+            case Nil => ""
+            case close => s" - did you mean ${util.StringUtil.oxford(close, "or")}?"
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/PhaseAssembly.scala:291]: mayte this doesn't carry its weight? can we do without?

File: src/compiler/scala/tools/nsc/PhaseAssembly.scala [hunk 9]
```diff
@@ -226,50 +232,92 @@ object DependencyGraph {
...
+      for (before <- p.runsBefore if before.nonEmpty)
+        if (checkConstraint(p)(before, "runsBefore", warn=warnAll)) addConstraint(p.phaseName, before, Follows)
+      // terminal follows every phase; parser is not before every phase because traverse uses edgeTo to find "anchors"
       if (p != end || p == end && p == start)
-        if (!p.runsBefore.exists(graph.components.contains))
-          graph.addEdge(p.phaseName, end.phaseName, Follows)
+        addConstraint(p.phaseName, end.phaseName, Follows)
+    }
+    if (warnAll)
+      dropped.foreach(p => warnings.addOne(s"Component ${p} dropped due to bad constraint"))
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/PhaseAssembly.scala:311]: this warning seems redundant, at least in the check files

File: src/compiler/scala/tools/nsc/PhaseAssembly.scala [hunk 10]
```diff
@@ -226,50 +232,92 @@ object DependencyGraph {
...
+    val purgedConstraints = constraints.filterInPlace {
+      case (from, to, w) => !dropped.contains(from) && !dropped.contains(to)
+    }.toList
+    new DependencyGraph(start.phaseName, phases.iterator.map(p => p.phaseName -> p).toMap).tap { graph =>
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/PhaseAssembly.scala:315]: what's the advantage of the new approach of collecting constraints before creating the `DependencyGraph`?

**Comment by som-snytt** [comment]
There are no more phase assembly errors.

File: src/compiler/scala/tools/nsc/PhaseAssembly.scala (hunk 10 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/PhaseAssembly.scala:315]: something about better errors, I guess.
